scripts/o3de/o3de/compatibility.py

```diff
@@ -0,0 +1,295 @@
+        return False
+
+    # compatibility will be based on the engine the project uses and the gems visible to
+    # the engine and project
+    compatible = gem_engine_compatible(gem_json_data, engine_json_data, gem_paths=gem_paths)
+    if compatible and check:
+        logger.info(f'{gem_json_data["gem_name"]} is compatible with this project and engine.')
+    elif not compatible:
+        return False
+
+    return True 
+
+
+def gem_engine_compatible(gem_json_data:dict, engine_json_data:dict, gem_paths:list = None, check:bool = False) -> bool:
+    """
+    Returns True if the gem is compatible with the indicated engine.
+    :param gem_json_data: gem json data dictionary
+    :param engine_json_data: engine json data dictionary
+    :param gem_paths: paths to all gems to use for dependency checks (optional)
+    :param check: if True always return True and ouput log info about the result
+    """
+    engine_compatible = o3de_object_is_compatible_with_engine(gem_json_data, engine_json_data)
+    if not engine_compatible:
+        error_msg = f'{gem_json_data["gem_name"]} is not known to be compatible with the '\
+            f'engine {engine_json_data["engine_name"]}{engine_json_data.get("version","")} '\
+             'and requires the --force parameter to enable.'
+        if check:
+            logger.info(error_msg)
+        else:
+            logger.error(error_msg)
+
+    if not gem_paths:
+        gem_paths = manifest.get_engine_gems()
+
+    return engine_compatible and gem_dependencies_compatible(gem_json_data, gem_paths=gem_paths, check=check)
+
+def o3de_object_is_compatible_with_engine(object_json_data:dict, engine_json_data:dict) -> bool:
+    """
+    Returns True if the engine is compatible with the provided object's compatible_engines and engine_api_versions information.
+    :param object_json_data: the json data for the object 
+    :param engine_json_data: the json data for the engine
+    """
+    if (object_json_data.get('compatible_engines') or object_json_data.get('engine_api_dependencies')):
+        return engine_is_compatible(
+            engine_json_data['engine_name'], engine_json_data.get('version',''), 
+            object_json_data.get('compatible_engines',''), object_json_data.get('engine_api_dependencies',''),
+            engine_json_data)
+    return True
+
+def engine_is_compatible(engine_name:str, engine_version:str, compatible_engines:list, engine_api_version_specifiers:list, engine_json_data:dict or None) -> bool:
+    """
+    Returns True if the engine is compatible with the provided compatible_engines and engine_api_versions information.
+    If a compatible_engine entry only has an engine name, it is assumed compatible with every engine version with that name.
+    If an engine_api_version entry only has an api name, it is assumed compatible with every engine api version with that name.
+    :param engine_name: the engine name
```

Comment:
Can you specify that if engine_name is `None`, the this engine containing the o3de scripts will be used?

Reply:
Or better yet, make the engine_json_data not optional and remove the `engine_name` and `engine_version` parameter.

There is no need for those parameters when the json data is supplied.